### PR TITLE
First attempt to introduce spmd_block in hpx

### PR DESCRIPTION
**Pull request by atrantan**
This PR tries to introduce code related to _spmd_block_ that was presented here http://stellar.cct.lsu.edu/pubs/extending_cpp_with_coarray_semantics_2016.pdf
A major difference in comparaison to what was presented in that paper is that _an image_ is no longer considered as a computation located in an _explicit_ locality. It is now a simple sequential computation able to cooperate with other ones.

### CODE REVIEW

**Comment by hkaiser** [comment]
@atrantan inspect is not happy with this: https://6417-4455628-gh.circle-artifacts.com/0/tmp/circle-artifacts.fNdJx9u/hpx_inspect_report.html

File: hpx/parallel/spmd_block.hpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        static_assert(std::is_same<spmd_block, first_type>::value,
+            "define_spmd_block() needs a lambda that " \
+            "has at least a spmd_block as 1st argument");
+
+        std::shared_ptr<hpx::lcos::local::barrier> barrier
+            = std::make_shared<hpx::lcos::local::barrier>(num_images);
+
+        return hpx::parallel::executor_traits<
+                typename std::decay<executor_type>::type
+            >::bulk_execute(
```
Comment (atrantan) [inline hpx/parallel/spmd_block.hpp]: Opt for synchronous `define_spmd_block()`. Done in 4d109e52d23bce52b3b418811faa88a294dd1953

File: hpx/parallel/spmd_block.hpp [hunk 2]
```diff
@@ -84,34 +87,47 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v2)
...
-            typename std::decay<F>::type f_;
+            std::string name_;
             std::size_t num_images_;
 
             template <typename ... Ts>
             void operator()(std::size_t image_id, Ts && ... ts) const
             {
-                spmd_block block(num_images_, image_id, barrier_);
+                hpx::lcos::barrier
+                    barrier(name_ + "_barrier" , num_images_, image_id);
```
Comment (atrantan) [inline hpx/parallel/spmd_block.hpp]: Not sure about how to correctly instanciate the barrier, but for now I've got a deadlock when `.wait()` is called

File: hpx/parallel/spmd_block.hpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                std::size_t num_images,
+                Args... args)
+            {
+                executor_type exec;
+                std::size_t offset = hpx::get_locality_id();
+                offset *= images_per_locality;
+
+                hpx::parallel::executor_traits<
+                    executor_type
+                    >::bulk_execute(
```
Comment (sithhell) [inline hpx/parallel/spmd_block.hpp]: Can't this be `return bulk_async_execute`?

Comment (atrantan) [inline hpx/parallel/spmd_block.hpp]: Sure. Done in 23ce7fe33d8afdc528631c85236d18d0e31bc94c

File: hpx/parallel/spmd_block.hpp (hunk 2 above)
Comment (atrantan) [inline hpx/parallel/spmd_block.hpp]: With commit 23ce7fe33d8afdc528631c85236d18d0e31bc94c. I'm now having an exception that give me more informations: terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<hpx::exception> >'
  what():  bind: Address already in use: HPX(network_error)``

File: hpx/parallel/spmd_block.hpp [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+                Args... args)
+            {
+                executor_type exec;
+                std::size_t offset = hpx::get_locality_id();
+                offset *= images_per_locality;
+
+                return
+                    hpx::parallel::executor_traits<
+                        executor_type
+                        >::bulk_execute(
```
Comment (sithhell) [inline hpx/parallel/spmd_block.hpp]: so why not `bulk_async_execute`?

Comment (atrantan) [inline hpx/parallel/spmd_block.hpp]: Maybe a good idea but for now I'm testing the synchronous version to make sure that everything is working fine. Got a deadlock when the `.wait()` `lcos::barrier` method is called.

**Comment by atrantan** [comment]
I'm now closing this PR due to issue related to `lcos::barrier`. For the next release, only _local_ `define_spmd_block ` will be submitted as release PR

**Pull request closed without merging**